build/org.eclipse.cdt.managedbuilder.core/src/org/eclipse/cdt/managedbuilder/core/jsoncdb/generator/CompilationDatabaseGenerator.java

```diff
@@ -0,0 +1,419 @@
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.LinkedHashSet;
+import java.util.List;
+
+import org.eclipse.cdt.core.settings.model.ICSourceEntry;
+import org.eclipse.cdt.core.settings.model.util.CDataUtil;
+import org.eclipse.cdt.managedbuilder.core.BuildException;
+import org.eclipse.cdt.managedbuilder.core.IConfiguration;
+import org.eclipse.cdt.managedbuilder.core.IFileInfo;
+import org.eclipse.cdt.managedbuilder.core.IFolderInfo;
+import org.eclipse.cdt.managedbuilder.core.IManagedCommandLineGenerator;
+import org.eclipse.cdt.managedbuilder.core.IManagedCommandLineInfo;
+import org.eclipse.cdt.managedbuilder.core.IOutputType;
+import org.eclipse.cdt.managedbuilder.core.IResourceInfo;
+import org.eclipse.cdt.managedbuilder.core.ITool;
+import org.eclipse.cdt.managedbuilder.core.ManagedBuildManager;
+import org.eclipse.cdt.managedbuilder.core.jsoncdb.CompilationDatabaseInformation;
+import org.eclipse.cdt.managedbuilder.core.jsoncdb.ICompilationDatabaseContributor;
+import org.eclipse.cdt.managedbuilder.internal.macros.FileContextData;
+import org.eclipse.cdt.managedbuilder.macros.BuildMacroException;
+import org.eclipse.cdt.managedbuilder.macros.IBuildMacroProvider;
+import org.eclipse.core.resources.IFile;
+import org.eclipse.core.resources.IFolder;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.IResource;
+import org.eclipse.core.resources.IResourceProxy;
+import org.eclipse.core.resources.IResourceProxyVisitor;
+import org.eclipse.core.resources.IResourceStatus;
+import org.eclipse.core.resources.IWorkspace;
+import org.eclipse.core.resources.ResourcesPlugin;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.Path;
+import org.eclipse.core.runtime.Platform;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.jdt.annotation.NonNull;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+public final class CompilationDatabaseGenerator {
+
+	private static final String CDB_FILENAME = "compile_commands.json"; //$NON-NLS-1$
+	private static final String BUILD_FOLDER = "build"; //$NON-NLS-1$
+	private static final String ERROR_MESSAGE = "Can not set contents to compile_commands.json file"; //$NON-NLS-1$
+
+	private IProject project;
+	private IConfiguration configuration;
+	private ICSourceEntry[] srcEntries;
+	private Collection<IResource> fileList;
+
+	public CompilationDatabaseGenerator(IProject proj, IConfiguration config) {
+		project = proj;
+		configuration = config;
+		srcEntries = config.getSourceEntries();
+	}
+
+	/**
+	 * @param proj
+	 *            Creates and populates compilation database file
+	 * @param config
+	 */
+	public void generate() {
+
+		IPath buildDirectory = project.getFullPath()
+				.append(IPath.SEPARATOR + CompilationDatabaseGenerator.BUILD_FOLDER);
```

Comment:
Why are we always generating into a new folder called "build" instead of into the config directory?

Reply:
ClangD expects to find the file on a folder called "build", or in the project root. We chose to add it to the "build" folder for Renesas, but I guess this can be discussed more.

Reply:
If we're generating the .clangd file into the project root then we could move the json file into the config directory, using the .clangd to point to it. However the ".clangd" file is expected to be checked into source control. If we write a config specific path into .clangd then it would be changed whenever the user changed build configs. That action wouldn't be expected to impact files in source control.

Reply:
I wonder which way around is better here. We can put `compile_commands.json` in `Debug/compile_commands.json` and `Release/compile_commands.json` and update `.clangd` when the active configuration changes. Or we can put in `build/compile_commands.json`, and require a build to switch between configurations.

My initial thought was to put it in Debug/Release, but in light of the conversation above (https://github.com/eclipse-cdt/cdt/pull/692#discussion_r1479172539) requiring a build all the time to ensure an updated compile_commands.json may be the best for now.

Therefore I will leave this as is and resolved for this PR and add a comment to #689

Reply:
@jonahgraham as my comment above. Updating .clangd on config changes is problematic as .clangd is supposed to be in source control according to clangd's documentation. Changing the active config would not be expected to impact source controlled files as it a user level preference.

Reply:
If I have understood correctly, there is one instance of clangd which will look for `.clangd` in ancestor folders of the source file and in `build` folders under each of those ancestor folders. In that case, for MBS projects only, could we update `.clangd` in the project root to point to the configuration databse in the active build configuration folder and mark the `.clangd` file as a _derived resource_?

Reply:
Conversation moved to https://github.com/eclipse-cdt/cdt/issues/689#issuecomment-1942583000 and marking this as resolved.